File: frontend/src/features/cache-probe/schemas.ts
import { z } from "zod";

export const CACHE_PROBE_ROLES = ["seed", "other"] as const;
export const CACHE_PROBE_CALL_STATUSES = ["hit", "miss", "error"] as const;
export const CACHE_PROBE_VERDICTS = [
  "cross_account_sharing",
  "no_cross_account_hit",
  "inconclusive",
] as const;

const CacheProbeRoleSchema = z.enum(CACHE_PROBE_ROLES);
const CacheProbeCallStatusSchema = z.enum(CACHE_PROBE_CALL_STATUSES);
const CacheProbeVerdictSchema = z.enum(CACHE_PROBE_VERDICTS);

export const CacheProbeAccountSchema = z.object({
  accountId: z.string().min(1),
  label: z.string().min(1),
});

export const CacheProbePressureSchema = z.object({
  underPressure: z.boolean(),
  reason: z.string().nullish(),
  detail: z.string().nullish(),
  selectableAccountCount: z.number().int().nonnegative(),
  eligibleAccountCount: z.number().int().nonnegative(),
  pressuredAccountCount: z.number().int().nonnegative(),
});

export const CacheProbePlanSchema = z.object({
  model: z.string().nullish(),
  seedAccount: CacheProbeAccountSchema.nullish(),
  availableOtherAccounts: z.array(CacheProbeAccountSchema).default([]),
  seedRepetitions: z.number().int().positive(),
  totalCalls: z.number().int().nonnegative(),
  estimatedInputTokensPerCall: z.number().int().nonnegative(),
  estimatedTotalInputTokens: z.number().int().nonnegative(),
  maxSeedRepetitions: z.number().int().positive(),
  maxOtherAccounts: z.number().int().positive(),
  pressure: CacheProbePressureSchema,
});

export const CacheProbeCallSchema = z.object({
  sequence: z.number().int().positive(),
  accountId: z.string().min(1),
  accountLabel: z.string().min(1),
  role: CacheProbeRoleSchema,
  status: CacheProbeCallStatusSchema,
  cacheHit: z.boolean(),
  inputTokens: z.number().int().nonnegative().nullish(),
  cachedTokens: z.number().int().nonnegative().nullish(),
  latencyMs: z.number().int().nonnegative(),
  errorCode: z.string().nullish(),
});

export const CacheProbeRunSchema = z.object({
  runId: z.string().min(1),
  model: z.string().min(1),
  startedAt: z.iso.datetime({ offset: true }),
  completedAt: z.iso.datetime({ offset: true }),
  seedAccount: CacheProbeAccountSchema,
  seedRepetitions: z.number().int().positive(),
  calls: z.array(CacheProbeCallSchema).default([]),
  seedHitCount: z.number().int().nonnegative(),
  seedCallCount: z.number().int().nonnegative(),
  otherHitCount: z.number().int().nonnegative(),
  otherCallCount: z.number().int().nonnegative(),
  crossAccountHit: z.boolean(),
  verdict: CacheProbeVerdictSchema,
});

export const CacheProbeRunRequestSchema = z.object({
  confirm: z.literal(true),
  seedRepetitions: z.number().int().positive(),
  otherAccountCount: z.number().int().positive(),
});

export type CacheProbeRole = z.infer<typeof CacheProbeRoleSchema>;
export type CacheProbeCallStatus = z.infer<typeof CacheProbeCallStatusSchema>;
export type CacheProbeVerdict = z.infer<typeof CacheProbeVerdictSchema>;
export type CacheProbeAccount = z.infer<typeof CacheProbeAccountSchema>;
export type CacheProbePressure = z.infer<typeof CacheProbePressureSchema>;
export type CacheProbePlan = z.infer<typeof CacheProbePlanSchema>;
export type CacheProbeCall = z.infer<typeof CacheProbeCallSchema>;
export type CacheProbeRun = z.infer<typeof CacheProbeRunSchema>;
export type CacheProbeRunRequest = z.infer<typeof CacheProbeRunRequestSchema>;

File: app/modules/cache_isolation_probe/schemas.py
"""Wire contract for the cross-account prompt cache isolation probe.

Only numbers, identifiers and enumerated statuses cross this boundary. The
generated prefix, the run nonce and every byte of upstream response content
stay inside the service.
"""

from __future__ import annotations

from datetime import datetime

from pydantic import Field

from app.modules.cache_isolation_probe.service import (
    DEFAULT_OTHER_ACCOUNTS,
    DEFAULT_SEED_REPETITIONS,
    MAX_OTHER_ACCOUNTS,
    MAX_SEED_REPETITIONS,
    ProbeCallRole,
    ProbeCallStatus,
    ProbeVerdict,
)
from app.modules.shared.schemas import DashboardModel


class CacheProbeAccountResponse(DashboardModel):
    account_id: str
    label: str


class CacheProbePressureResponse(DashboardModel):
    under_pressure: bool
    reason: str | None = None
    detail: str | None = None
    selectable_account_count: int
    eligible_account_count: int
    pressured_account_count: int


class CacheProbePlanResponse(DashboardModel):
    model: str | None = None
    seed_account: CacheProbeAccountResponse | None = None
    #: Every sibling the pool can offer, capped; the client picks how many of
    #: them a run should use.
    available_other_accounts: list[CacheProbeAccountResponse] = Field(default_factory=list)
    seed_repetitions: int
    total_calls: int
    estimated_input_tokens_per_call: int
    estimated_total_input_tokens: int
    max_seed_repetitions: int
    max_other_accounts: int
    pressure: CacheProbePressureResponse


class CacheProbeRunRequest(DashboardModel):
    #: Explicit, typed acknowledgement that the run spends real quota. The
    #: endpoint refuses anything but ``true`` so a bare POST cannot bill the
    #: pool by accident.
    confirm: bool = False
    model: str | None = None
    seed_repetitions: int = Field(default=DEFAULT_SEED_REPETITIONS, ge=1, le=MAX_SEED_REPETITIONS)
    #: Upper bound on sibling accounts, not a demand: a smaller pool still
    #: runs, with fewer rows. The plan response lists what will actually be
    #: called.
    other_account_count: int = Field(default=DEFAULT_OTHER_ACCOUNTS, ge=1, le=MAX_OTHER_ACCOUNTS)


class CacheProbeCallResponse(DashboardModel):
    sequence: int
    account_id: str
    account_label: str
    role: ProbeCallRole
    status: ProbeCallStatus
    cache_hit: bool
    input_tokens: int | None = None
    cached_tokens: int | None = None
    latency_ms: int
    error_code: str | None = None


class CacheProbeRunResponse(DashboardModel):
    run_id: str
    model: str
    started_at: datetime
    completed_at: datetime
    seed_account: CacheProbeAccountResponse
    seed_repetitions: int
    calls: list[CacheProbeCallResponse] = Field(default_factory=list)
    seed_hit_count: int
    seed_call_count: int
    other_hit_count: int
    other_call_count: int
    cross_account_hit: bool
    verdict: ProbeVerdict
